src/mlpack/methods/matrix_completion/mc_fw_update_matrix.hpp

```diff
@@ -0,0 +1,238 @@
+   * faster.
+   *
+   * See Algorithm 3 of
+   * @code
+   * Rao N, Shah P, Wright S  (2015)
+   * Forward--backward greedy algorithms for atomic norm regularization.
+   * IEEE Transactions on Signal Processing 63(21):5798–5811
+   * @endcode
+   *
+   * @param F threshold number.
+   * @param function function to be optimized.
+   */
+  void Truncation(const double F, MatrixCompletionFWFunction& function)
+  {
+    arma::mat X;
+    RecoverMatrix(X);
+
+    arma::mat U, V;
+    arma::vec s;
+    arma::uword k;
+    if (!arma::svd_econ(U, s, V, X))
+      Log::Fatal << "Truncation: armadillo svd_econ() failed!" << std::endl;
+    size_t r = s.n_elem;
+
+    for (size_t i = 1; i < r; i++)
+    {
+      // Try deleting the matrix atom with smallest coefficient.
+      s.min(k);
+      X = X - s(k) * U.col(k) * arma::trans(V.col(k));
+
+      // Cannot delete atom, just break.
+      if ((function.Evaluate(X) > F))
+        break;
+
+      // Delete this atom.
+      U.shed_col(k);
+      V.shed_col(k);
+      s.shed_row(k);
+    } // for
+
+    // Redefine the atoms.
+    currentAtoms.set_size(U.n_rows, V.n_rows, s.n_rows);
+    currentCoeffs = s;
+    for (arma::uword i = 0; i < s.n_rows; i++)
+      currentAtoms.slice(i) = U.col(i) * arma::trans(V.col(i));
+  } // Truncation()
+
+  /**
+   * Enhancement step. Recalculate the coefficients of the atoms. That is,
+   * try to get smaller object function value with updating the singular values
+   * only, without changing the left and right singular vectors. In order to
+   * impose the nuclear norm constraint, projected gradient method is used.
+   *
+   * Notice that for matrix completion problem, the object function has Lipchitz
+   * constant 1. So theoretically the step size for projected gradient method
+   * could be anything smaller than 1.
+   *
+   * See algorithm 1 of the above mentioned paper.
+   */
+  void ProjectedGradientEnhancement(MatrixCompletionFWFunction& function,
+                                    double stepSize = 0.9,
+                                    size_t maxIterations = 100,
+                                    double tolerance = 1e-3)
+  {
+    arma::mat x;
+    RecoverMatrix(x);
+    double value = function.Evaluate(x);
+
+    for (size_t iter = 1; iter < maxIterations; iter++)
+    {
+      // Update currentCoeffs with gradient descent method.
+      // We caculate the gradient directly here, so we don't need to generate
+      // a large sparse matrix.
+      arma::vec entries;
+      function.GetKnownEntries(x, entries);
+      entries = entries - function.Values();
+
+      arma::vec gradient(currentCoeffs.n_elem);
+      for (size_t i = 0; i < gradient.n_elem; i++)
+      {
+        arma::vec entriesAtom;
+        function.GetKnownEntries(currentAtoms.slice(i), entriesAtom);
+
+        gradient(i) = arma::dot(entriesAtom, entries);
+      }
+      currentCoeffs = currentCoeffs - stepSize * gradient;
```

Comment:
Just a minor note; you could just write: ``currentCoeffs -= function.Values();``.